src/foam/nanos/auth/test/ServiceProviderAwareTest.js

```diff
@@ -0,0 +1,250 @@
+
+         // permission setup
+        DAO groupPermissionJunctionDAO = (DAO) x.get("localGroupPermissionJunctionDAO");
+        groupPermissionJunctionDAO.put(new GroupPermissionJunction.Builder(x).setSourceId("test").setTargetId("user.read.*").build());
+        groupPermissionJunctionDAO.put(new GroupPermissionJunction.Builder(x).setSourceId("fail").setTargetId("user.read.*").build());
+        groupPermissionJunctionDAO.put(new GroupPermissionJunction.Builder(x).setSourceId("test").setTargetId("spid.read.test").build());
+        groupPermissionJunctionDAO.put(new GroupPermissionJunction.Builder(x).setSourceId("fail").setTargetId("spid.read.fail").build());
+
+        foam.nanos.app.AppConfig appConfig = (foam.nanos.app.AppConfig) x.get("appConfig");
+        appConfig.setDefaultSpid("test");
+        X y = x.put("appConfig", appConfig);
+
+        User user = new User.Builder(x)
+          .setId(99995)
+          .setEmail("test@example.com")
+          .setSpid("test")
+          .setGroup("admin")
+          .build();
+        y = Auth.sudo(y, user);
+
+        testAware(y);
+        testReference(y);
+      `
+    },
+    {
+      name: 'testAware',
+      args: [
+        { type: 'Context', name: 'x' }
+      ],
+      javaCode: `
+        AuthService auth = (AuthService) x.get("auth");
+
+        User ctxUser = (User) x.get("user");
+        DAO delegate = new MDAO(User.getOwnClassInfo());
+        DAO dao = (DAO) new ServiceProviderAwareDAO.Builder(x).setDelegate(delegate).build();
+        ctxUser = (User) delegate.put_(x, ctxUser);
+        int idx = 0;
+
+        test("test".equals(ctxUser.getSpid()), "["+(idx++)+"] context user spid is set");
+        User user = (User) new User.Builder(x)
+          .setId(99999)
+          .setFirstName("first")
+          .setLastName("last")
+          .setEmail("99999@test.com")
+          .setGroup("test")
+          .build();
+        user = (User) dao.put(user);
+        test(ctxUser.getSpid().equals(user.getSpid()), "["+(idx++)+"] ServiceProviderAwareDAO put sets spid. ");
+        user = (User) dao.find(user.getId());
+        test(user != null, "["+(idx++)+"] ServiceProviderAwareDAO find matches spid.");
+
+        User user2 = (User) new User.Builder(x)
+          .setId(99998)
+          .setFirstName("first_two")
+          .setLastName("last_two")
+          .setEmail("99998@test.com")
+          .setGroup("test")
+          .build();
+        user2 = (User) dao.put(user2);
+
+        ArraySink sink = new ArraySink();
+        dao.select(sink);
+        List users = sink.getArray();
+        test (users.size() == 3, "["+(idx++)+"]  ServiceProvicerAwareDAO select matched on spid. expected: 3, found: "+users.size());
+
+        ctxUser = new User.Builder(x)
+          .setId(99996)
+          .setEmail("test@example.com")
+          .setSpid("fail")
+          .setGroup("admin")
+          .build();
+        ctxUser = (User) delegate.put_(x, ctxUser);
```

Comment:
Seems like the context user (admin) who belongs to "test" spid before can add new user for other spid "fail". Is it intended or I misunderstand the flow here?

Reply:
The delegate is an explicit MDAO - not going through the normal DAO stack.  Just for test setup.